### ISSUE TITLE
Twisted/GTK programs don't close properly on Ubuntu 12.04

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5503]
|<img alt="alkisg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| alkisg reported|
|-|-|
|Trac ID|trac#5503|
|Type|release blocker: regression|
|Created|2012-02-29 21:22:45Z|

I've been experiencing this with our www.epoptes.org application, but I've been able to reproduce the problem with http://twistedmatrix.com/documents/current/core/examples/pbgtk2.py too.
I.e. when I run one of those apps from a console, and then close the application window, the window does close, but my console prompt doesn't return. Sometimes the app closes after 20-30 seconds, sometimes it just sits there forever waiting. An strace tells me it's waiting at some poll() call.

The same programs run fine in previous versions of Ubuntu.
Also, if I put a `reactor.crash()` just before `reactor.stop()`, then those programs terminate immediately.

<details><summary>Searchable metadata</summary>

```
trac-id__5503 5503
type__release_blocker__regression release blocker: regression
reporter__alkisg alkisg
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1330550565000000 1330550565000000
changetime__1334077724000000 1334077724000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_1"></a>|@thijstriemstra commented|
|-|-|

Replying to [alkisg](#5503):
> I've been experiencing this with our www.epoptes.org application, but I've been able to reproduce the problem with http://twistedmatrix.com/documents/current/core/examples/pbgtk2.py too.
> I.e. when I run one of those apps from a console, and then close the application window, the window does close, but my console prompt doesn't return.

When I run this example and close the application window I get the following traceback (ubuntu 11.10):

```python
$ python doc/core/examples/pbgtk2.py 
Traceback (most recent call last):
  File "doc/core/examples/pbgtk2.py", line 46, in on_loginDialog_response
    handlers.get(response)()
  File "doc/core/examples/pbgtk2.py", line 58, in windowClosed
    self.deferredResult.errback()
  File "/home/thijs/workspaces/opensource/Twisted/trunk/twisted/internet/defer.py", line 393, in errback
    fail = failure.Failure(captureVars=self.debug)
  File "/home/thijs/workspaces/opensource/Twisted/trunk/twisted/python/failure.py", line 217, in __init__
    raise NoCurrentExceptionError()
twisted.python.failure.NoCurrentExceptionError
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

> When I run this example and close the application window I get the following traceback (ubuntu 11.10): 

That's an unrelated bug in the example.  If you actually *connect* to a PB server first then you'll experience the reported problem.

**Comment by twisted-trac** [comment]
|<img alt="alkisg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|alkisg set status to `closed`|
|-|-|

Problem appears to be fixed in package python-twisted-core 11.1.0-1ubuntu2:
  * debian/patches/01_posix_wakeups.patch:
    - Backport wakeup change from trunk gtk2refactoring (LP: #935756)
